crates/sui-graphql-rpc/src/types/available_range.rs

```diff
@@ -1,28 +1,92 @@
 pub(crate) struct AvailableRange {
     pub first: u64,
     pub last: u64,
 }
 
-// TODO: do both in one query?
 /// Range of checkpoints that the RPC is guaranteed to produce a consistent response for.
 #[Object]
 impl AvailableRange {
     async fn first(&self, ctx: &Context<'_>) -> Result<Option<Checkpoint>> {
-        Checkpoint::query(ctx, CheckpointId::by_seq_num(self.first), Some(self.last))
+        Checkpoint::query(ctx, CheckpointId::by_seq_num(self.first), self.last)
             .await
             .extend()
     }
 
     async fn last(&self, ctx: &Context<'_>) -> Result<Option<Checkpoint>> {
-        Checkpoint::query(ctx, CheckpointId::by_seq_num(self.last), Some(self.last))
+        Checkpoint::query(ctx, CheckpointId::by_seq_num(self.last), self.last)
             .await
             .extend()
     }
 }
+
+impl AvailableRange {
+    /// Look up the available range when viewing the data consistently at `checkpoint_viewed_at`.
+    pub(crate) async fn query(db: &Db, checkpoint_viewed_at: u64) -> Result<Self, Error> {
+        let Some(range): Option<Self> = db
+            .execute(move |conn| Self::result(conn, checkpoint_viewed_at))
+            .await
+            .map_err(|e| Error::Internal(format!("Failed to fetch available range: {e}")))?
+        else {
+            return Err(Error::Client(format!(
+                "Requesting data at checkpoint {checkpoint_viewed_at}, outside the available \
+                 range.",
+            )));
+        };
+
+        Ok(range)
+    }
+
+    /// Look up the available range when viewing the data consistently at `checkpoint_viewed_at`.
+    /// Made available on the `Conn` type to make it easier to call as part of other queries.
+    ///
+    /// Returns an error if there was an issue querying the database, Ok(None) if the checkpoint
+    /// being viewed is not in the database's available range, or Ok(Some(AvailableRange))
+    /// otherwise.
+    pub(crate) fn result(conn: &mut Conn, checkpoint_viewed_at: u64) -> QueryResult<Option<Self>> {
+        use checkpoints::dsl as checkpoints;
+        use objects_snapshot::dsl as snapshots;
+
+        let checkpoint_range: Vec<i64> = conn.results(move || {
+            let rhs = checkpoints::checkpoints
+                .select(checkpoints::sequence_number)
+                .order(checkpoints::sequence_number.desc())
+                .limit(1);
+
+            let lhs = snapshots::objects_snapshot
+                .select(snapshots::checkpoint_sequence_number)
+                .order(snapshots::checkpoint_sequence_number.desc())
+                .limit(1);
+
+            lhs.union(rhs)
+        })?;
+
+        let (first, mut last) = match checkpoint_range.as_slice() {
+            [] => (0, 0),
+            [single_value] => (0, *single_value as u64),
+            values => {
+                let min_value = *values.iter().min().unwrap();
+                let max_value = *values.iter().max().unwrap();
```

Comment:
I assume these `unwrap` are `safe` because the iterator shouldn't be empty, given the pattern match , right?